<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>FQTL • fqtl</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="FQTL">
<meta property="og:description" content="Factored QTL estimation">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">fqtl</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.4.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="a-factorization-approach-to-multi-tissue-multi-snp-qtl-modeling">A factorization approach to multi-tissue multi-SNP QTL modeling<a class="anchor" aria-label="anchor" href="#a-factorization-approach-to-multi-tissue-multi-snp-qtl-modeling"></a>
</h1></div>
<p>NOTE: We are restructuring analysis pipeline and documents. Results and codes are temporary.</p>
<div class="section level2">
<h2 id="contributors">Contributors<a class="anchor" aria-label="anchor" href="#contributors"></a>
</h2>
<p>Methods are designed by Yongjin Park, Abhishek Sarkar, Kunal Bhutani, and Manolis Kellis; maintained by Yongjin Park (<code>ypp@csail.mit.edu</code>).</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install R package:</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"ypark/fqtl"</span><span class="op">)</span></span></code></pre>
<p>To successfully compile the Rcpp code, <code>R</code> must have been compiled with a compiler which supports C++14. Make sure your R development environment support <code>C++14</code> by including <code>-std=c++14</code> to <code>CFLAGS</code> and <code>CXXFLAGS</code> in <code>~/.R/Makevars</code> file. For instance,</p>
<pre><code>CXX = g++-6
CXX14 = g++-6
CXXFLAGS = -O3 -std=c++14
CFLAGS = -O3 -std=c++14
PKG_LDFLAGS = -Wl,--no-as-needed</code></pre>
<p>To speed up matrix-vector multiplication and vectorized random number generation, one can compile the package with <em>Intel MKL</em> or <em>openblas</em>.</p>
</div>
<div class="section level2">
<h2 id="results">Results<a class="anchor" aria-label="anchor" href="#results"></a>
</h2>
<p>Analysis pipeline and pre-calculated effect sizes computed in 44 GTEx tissues (v6p) can be found in separate repository.</p>
<p><a href="https://github.mit.edu/ypp/fqtl-gtex" class="external-link">fqtl-gtex</a></p>
</div>
<div class="section level2">
<h2 id="models">Models<a class="anchor" aria-label="anchor" href="#models"></a>
</h2>
<div class="section level3">
<h3 id="gaussian-model">Gaussian model<a class="anchor" aria-label="anchor" href="#gaussian-model"></a>
</h3>
<p>In Park and Sarkar <em>et al.</em> we define mean, variance and log-likelihood:</p>
<pre><code>E[y] = η1,
V[y] = Vmin + (Vmax - Vmin) * sigmoid(η2)
L[y] = -0.5 * (y - E[y])^2 / V[y] - 0.5 * ln V[y]</code></pre>
<p>Here we stabilized error variance, restricting on [<code>Vmin</code>, <code>Vmax</code>], with <code>Vmax = observed V[y]</code> and <code>Vmin = 1e-4 * Vmax</code>. A new version of the paper will soon be posted.</p>
</div>
<div class="section level3">
<h3 id="negative-binomial-model">Negative binomial model<a class="anchor" aria-label="anchor" href="#negative-binomial-model"></a>
</h3>
<p>Robinson and Smyth (2008), Anders and Huber (2010) popularized negative binomial models for analysis of high-throughput sequencing data.</p>
<pre><code>b[y] = exp(- η1)
E[y] = exp(Xθ + ln (1 + 1/φ))
V[y] = E[y] * (1 + E[y] / (1 + 1/φ))

L[y] = ln Gam(y + 1/φ + 1) - ln Gam(1/φ + 1)
       - y * ln(1 + exp(- η1))
       - (1/φ + 1) * ln(1 + exp(η1))</code></pre>
<p>Like Gaussian model, we could bound variance model <code>V[y] &lt; observed V[y]</code>. Probably most loosened model fit to the data is sample mean <code>Ybar</code>. If there were no association of covariates, sample mean would be unbiased estimation of rate parameter of the underlying Poisson distribution.</p>
<pre><code>worst variance = Ybar * (1 + Ybar /(1 + 1/φ))
               &lt; Ybar * (1 + Ybar * φ)
               &lt; Vobs</code></pre>
<p>Therefore we can claim that</p>
<pre><code><span><span class="va">φmax</span> <span class="op">=</span> <span class="op">(</span><span class="va">Vobs</span> <span class="op">/</span> <span class="va">Ybar</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span> <span class="op">/</span> <span class="va">Ybar</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="negative-binomial-model-redefined-with-psuedo-count">Negative binomial model (redefined with psuedo-count)<a class="anchor" aria-label="anchor" href="#negative-binomial-model-redefined-with-psuedo-count"></a>
</h3>
<p>We define negative binomial model adding pseudo-counts.</p>
<pre><code>p(y|α,β) = Γ(α + υ + α0)/[Γ(y+1)Γ(α + α0)] (1 + 1/β)^-(α + α0) (1 + β)^-y</code></pre>
<p>where</p>
<pre><code>E[y|α,α0,β] = (α + α0)/β

V[y|α,α0,β] = (α + α0)β + (α + α0)β^{2}

μ = (α0 + 1/φ) / β = exp[ η{μ} + ln (α0 + 1/φ) ]

σ^2 = μ + μ^{2} / (α0 + α)

α = αmin + (αmax - αmin) sigmoid(- η(nu))

β = exp(-η(μ))</code></pre>
</div>
<div class="section level3">
<h3 id="voom-model">Voom model<a class="anchor" aria-label="anchor" href="#voom-model"></a>
</h3>
<p>Law <em>et al.</em> (2014) presents log2-transformed over-dispersion model. They derived mean-variance relationship of the transformed random variable by the Delta method. We reparameterized the model as follows.</p>
<pre><code>E[y] = η1,
V[y] = exp(- η1) + φ
φ  = φmin + (φmax - φmin) * sigmoid(η2)</code></pre>
</div>
</div>
<div class="section level2">
<h2 id="statistical-inference">Statistical inference<a class="anchor" aria-label="anchor" href="#statistical-inference"></a>
</h2>
<p>We developed variational inference algorithm based on stochastic gradient algorithm (Paisley <em>et al.</em> 2012). Especially we tuned the method to better accommodate high-dimensional generalized linear models with data-driven sparsity.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Park, Y., Sarkar, A. K., Bhutani, K., &amp; Kellis, M. (2017). Multi-tissue polygenic models for transcriptome-wide association studies. bioRxiv, 107623. <a href="http://doi.org/10.1101/107623" class="external-link uri">http://doi.org/10.1101/107623</a></p>
<p>Robinson, M. D., &amp; Smyth, G. K. (2008). Small-sample estimation of negative binomial dispersion, with applications to SAGE data. Biostatistics (Oxford, England), 9(2), 321–332. <a href="http://doi.org/10.1093/biostatistics/kxm030" class="external-link uri">http://doi.org/10.1093/biostatistics/kxm030</a></p>
<p>Anders, S., &amp; Huber, W. (2010). Differential expression analysis for sequence count data. Genome Biology, 11(10), R106. <a href="http://doi.org/10.1186/gb-2010-11-10-r106" class="external-link uri">http://doi.org/10.1186/gb-2010-11-10-r106</a></p>
<p>Law, C. W., Chen, Y., Shi, W., &amp; Smyth, G. K. (2014). voom: Precision weights unlock linear model analysis tools for RNA-seq read counts. Genome Biology, 15(2), R29.</p>
<p>Paisley, J., Blei, D., &amp; Jordan, M. (2012). Variational Bayesian Inference with Stochastic Search. In J. Langford &amp; J. Pineau (Eds.), (pp. 1367–1374). Presented at the Proceedings of the 28th International Conference on Machine Learning, New York, NY, USA: Omnipress.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE-text.html">LICENSE</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing fqtl</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Yongjin Park <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Yongjin Park.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
